autogpts/autogpt/autogpt/agents/agent.py

```diff
@@ -64,157 +73,181 @@ class AgentConfiguration(BaseAgentConfiguration):
 
-        if include_os_info is None:
-            include_os_info = self.legacy_config.execute_local_commands
+        # Get commands
+        self.commands: list[Command] = list(
+            await self.foreach_components("get_commands")
+        )
 
-        return super().build_prompt(
-            *args,
-            extra_messages=extra_messages,
-            include_os_info=include_os_info,
-            **kwargs,
+        # Get messages
+        messages: list[ChatMessage] = list(
+            await self.foreach_components("get_messages")
         )
 
-    def on_before_think(self, *args, **kwargs) -> ChatPrompt:
-        prompt = super().on_before_think(*args, **kwargs)
+        prompt: ChatPrompt = await self.foreach_components(
+            "build_prompt",
+            messages,
+            self.commands,
+            self.state.task,
+            self.state.ai_profile,
+            directives,
+        )
 
         self.log_cycle_handler.log_count_within_cycle = 0
         self.log_cycle_handler.log_cycle(
-            self.ai_profile.ai_name,
+            self.state.ai_profile.ai_name,
             self.created_at,
             self.config.cycle_count,
             prompt.raw(),
             CURRENT_CONTEXT_FILE_NAME,
         )
-        return prompt
-
-    def parse_and_process_response(
-        self, llm_response: AssistantChatMessage, *args, **kwargs
-    ) -> Agent.ThoughtProcessOutput:
-        for plugin in self.config.plugins:
-            if not plugin.can_handle_post_planning():
-                continue
-            llm_response.content = plugin.post_planning(llm_response.content or "")
-
-        (
-            command_name,
-            arguments,
-            assistant_reply_dict,
-        ) = self.prompt_strategy.parse_response_content(llm_response)
-
-        # Check if command_name and arguments are already in the event_history
-        if self.event_history.matches_last_command(command_name, arguments):
-            raise DuplicateOperationError(
-                f"The command {command_name} with arguments {arguments} "
-                f"has been just executed."
+
+        logger.debug(f"Executing prompt:\n{dump_prompt(prompt)}")
+        output = await self.complete_and_parse(prompt)
+        self.config.cycle_count += 1
+
+        return output
+
+    @retry()
+    async def complete_and_parse(
+        self, prompt: ChatPrompt, exception: Optional[Exception] = None
+    ) -> ThoughtProcessOutput:
+        if exception:
+            prompt.messages.append(ChatMessage.system(f"Error: {exception}"))
+
+        response: AssistantChatMessage = (
+            await self.llm_provider.create_chat_completion_raw(
+                prompt.messages,
+                functions=(
+                    get_openai_command_specs(self.commands)
+                    if self.config.use_functions_api
+                    else []
+                ),
+                model_name=self.llm.name,
             )
+        )
```

Comment:
I moved response parsing outside of llm provider since it isn't the provider responsibility and now it's more linear (instead of passing parse function to llm provider). If parsing fails entire `complete_and_parse` will restart (due to `retry` decorator)